### PR TITLE
Fix global mode

### PR DESCRIPTION
**Pull request by miparnisari**
## Description

@elbuo8 will add more details later :) 

## References

Closes https://github.com/mailgun/gubernator/issues/208

### CODE REVIEW

File: global.go [hunk 1]
```diff
@@ -157,6 +160,7 @@ func (gm *globalManager) sendHits(hits map[string]*RateLimitReq) {
 		fan.Run(func(in interface{}) error {
 			p := in.(*pair)
 			ctx, cancel := context.WithTimeout(context.Background(), gm.conf.GlobalTimeout)
+			gm.log.Infof("calling owner of key: %s with hits: %d", p.req.Requests[0].UniqueKey, p.req.Requests[0].Hits)
```
Comment (elbuo8) [inline global.go:163]:
```suggestion

```

File: global.go [hunk 2]
```diff
@@ -247,6 +234,7 @@ func (gm *globalManager) broadcastPeers(ctx context.Context, updates map[string]
 		fan.Run(func(in interface{}) error {
 			peer := in.(*PeerClient)
 			ctx, cancel := context.WithTimeout(ctx, gm.conf.GlobalTimeout)
+			gm.log.Infof("calling peer of key: %s with hits: %d", req.Globals[0].Key, req.Globals[0].Status.Remaining)
```
Comment (elbuo8) [inline global.go:237]:
```suggestion

```

File: gubernator.go [hunk 3]
```diff
@@ -427,13 +429,29 @@ func (s *V1Instance) getGlobalRateLimit(ctx context.Context, req *RateLimitReq)
 // UpdatePeerGlobals updates the local cache with a list of global rate limits. This method should only
 // be called by a peer who is the owner of a global rate limit.
 func (s *V1Instance) UpdatePeerGlobals(ctx context.Context, r *UpdatePeerGlobalsReq) (*UpdatePeerGlobalsResp, error) {
+	now := MillisecondNow()
 	for _, g := range r.Globals {
 		item := &CacheItem{
-			ExpireAt:  g.Status.ResetTime,
+			ExpireAt:  g.Status.ResetTime + 100000,
```
Comment (elbuo8) [inline gubernator.go:435]:
```suggestion
			ExpireAt:  g.Status.ResetTime + 1000, // account for clock drift from owner where `ResetTime` might already be less than current time of the local machine. 
```

This value is arbitrary, perhaps should be a config or a constant. This prevents an immediate cache miss in small windows where clock drift might happen.

File: algorithms.go [hunk 4]
```diff
@@ -383,16 +383,17 @@ func leakyBucket(ctx context.Context, s Store, c Cache, r *RateLimitReq) (resp *
...
 			rl.ResetTime = now + (rl.Limit-rl.Remaining)*int64(rate)
 			return rl, nil
 		}
 
-		// If requested is more than available, then return over the limit
-		// without updating the bucket.
+		// If requested is more than available, drain bucket in order to converge as everything is returning OVER_LIMIT.
 		if r.Hits > int64(b.Remaining) {
 			metricOverLimitCounter.Add(1)
+			b.Remaining = 0
```
Comment (elbuo8) [inline algorithms.go:395]: This was changed because the owner receives collapsed/accumulated hits, so if the remainder is 50, drip/leak brings it to 75 and the peers reported a total of 200 hits, it's fair to say that the quota has been consumed AND that the tokens be taken away given that drip was applied.

File: gubernator.go [hunk 5]
```diff
@@ -396,21 +396,23 @@ func (s *V1Instance) getGlobalRateLimit(ctx context.Context, req *RateLimitReq)
...
-	if err != nil {
-		countError(err, "Error in workerPool.GetCacheItem")
-		return nil, errors.Wrap(err, "during in workerPool.GetCacheItem")
-	}
-	if ok {
-		// Global rate limits are always stored as RateLimitResp regardless of algorithm
-		rl, ok := item.Value.(*RateLimitResp)
-		if ok {
-			return rl, nil
+	/*
```
Comment (elbuo8) [inline gubernator.go:399]: > However, for LEAKY_BUCKET the peer doesn't really know what tokens have leaked since we last checked.... I guess we could have the peer calculate the leak... but that sounds icky... Maybe someone can convince me otherwise?

@thrawn01 there is more to this story we'd like to discuss.

This change (left in comments so its easier to discuss, we can remove or uncomment later) basically applies drip on both peers and owner. Thus the owner has the true count while peers are constantly changing.

This has benefits and drawbacks.

Benefits:
1. If owner takes too long to broadcast (network failure, saturation, etc) the peers can eventually converge on OVER_LIMIT and the system can protect itself.
2. If owner takes too long to broadcast, peers can start letting traffic through instead of completely closing the circuit (stuck on OVER_LIMIT until TTL of the cached response).
3. If owner takes too long to broadcast, peers can start letting traffic through slowly given that the limit won't be immediately reset once the TTL of the cached response happens.
4. If peer(s) suffer from network issues where their hits can't be flushed, they can operate individually until the converge. If high rate, they will converge at rate limiting, otherwise they will constantly be flapping.

Basically we want to reduce the single point of failure the owner introduces by being the only one that can lower the remaining count. Specially if the Duration is 1s (or less) with a high leak/drip rate.
This comes with a particular drawback. It makes it so that you give away more tokens that you'd normally would. For example, if you set an RPS of 100 and you have 20 instances, that means that you will have 20*100 RPS in the case the owner is not fast enough. The more nodes you add, the more permissive it is because it's constantly dripping on each individual node. Thus you have to be mindful of your underlying capacity/fleet of gubernator. The smaller the fleet, the more strict it will be, the bigger, the more permissive it will be in high drip/leak scenarios.

File: algorithms.go (hunk 4 above)
Comment (thrawn01) [reply algorithms.go:395]: This is a great catch! However, this changes the semantics of asking for more than what is available. If we went forward with this change, any request that asks for more than is available will gobble up the remainder without actually using them.

For example: If my email limit is 1,000 and I've already sent 500 emails. Then I request to send 1,000 more emails, I would expect the 1,000 email request to be rejected as "over the limit". However, subsequent requests for 500 emails or less should be allowed, as I have 500 emails remaining within my rate limit window.

I think I have an idea of how to solve this. I'm adding some tests to cover this scenario and will have a PR soon.

Comment (Baliedge) [reply algorithms.go:395]: I had just merged in this behavior as a new behavior type `DRAIN_OVER_LIMIT`.  I faced the same usability issue, but I did not want to break legacy applications that may have depended on old logic.
https://github.com/mailgun/gubernator/pull/209

**Comment by thrawn01** [comment]
I found a few other issues in this PR, and embraced your feed back. I think you are on the right track. I've created a new PR with the commits from this PR. #219 

Please provide feed back on these changes, and I hope to work with @Baliedge next week to get this and all your other PR's merged!

**Comment by Baliedge** [comment]
Closing in favor of: https://github.com/mailgun/gubernator/pull/219
